src/LsCommand.cc

```diff
@@ -0,0 +1,253 @@
+  DIR* dir = opendir(dir_name.c_str());
+  if (!dir) {
+    cerr << "Cannot open " << dir_name << endl;
+    return false;
+  }
+
+  size_t bytes = 0;
+  while (struct dirent* ent = readdir(dir)) {
+    string path = dir_name + "/" + ent->d_name;
+
+    struct stat st;
+    if (stat(path.c_str(), &st) == -1) {
+      cerr << "stat " << path << " failed\n";
+      return false;
+    }
+
+    bytes += st.st_size;
+  }
+
+  const char suffixes[] = " KMGT";
+  double size = bytes;
+  size_t suffix_idx = 0;
+  while (size >= 1000.0) {
+    size /= 1024.0;
+    suffix_idx++;
+  }
+  char suffix = suffixes[suffix_idx];
+
+  ostringstream cvt;
+
+  if (suffix == ' ') {
+    cvt << bytes;
+  } else if (size >= 10) {
+    cvt << int(size) << suffix;
+  } else {
+    cvt << fixed << setprecision(1) << size << suffix;
+  }
+
+  size_str = cvt.str();
+  return true;
+}
+
+static bool is_valid_trace(const string& entry) {
+  if (entry[0] == '.')
+    return false;
+  if (entry[0] == '#')
+    return false;
+  if (entry[entry.length() - 1] == '~')
+    return false;
+  return true;
+}
+
+static string get_exec_path(TraceInfo& info) {
+  while (true) {
+    TraceTaskEvent r = info.second->read_task_event();
+    if (r.type() == TraceTaskEvent::NONE) {
+      break;
+    }
+    if (r.type() == TraceTaskEvent::EXEC) {
+      return r.cmd_line()[0];
+    }
+  }
+  return string();
+}
+
+static int ls(const string& traces_dir, const LsFlags& flags) {
+  DIR* dir = opendir(traces_dir.c_str());
+  if (!dir) {
+    fprintf(stdout, "Cannot open %s", traces_dir.c_str());
+    return 1;
+  }
+
+  vector<TraceInfo> traces;
+
+  while (struct dirent* trace_dir = readdir(dir)) {
+    if (!is_valid_trace(trace_dir->d_name)) {
+      continue;
+    }
+    string full_trace_dir = traces_dir + "/" + trace_dir->d_name;
+    traces.emplace_back(trace_dir, new TraceReader(full_trace_dir));
+  }
+
+  sort(traces.begin(), traces.end(),
+       flags.sort_by_time ? compare_by_time : compare_by_name);
+
+  if (flags.reverse) {
+    reverse(begin(traces), end(traces));
+  };
+
+  if (!flags.full_listing) {
+    for (TraceInfo t : traces) {
+      cout << t.first->d_name << "\n";
+    }
+    return 0;
+  }
+
+  int max_name_size =
+    accumulate(traces.begin(), traces.end(), 0, [](int m, TraceInfo t) {
+        return max(m, static_cast<int>(strlen(t.first->d_name)));
+    });
+
+  fprintf(stdout, "%-*s %19s %6s %5s %s\n", max_name_size,
+          "NAME", "WHEN", "DUR", "SIZE", "CMD");
+
+  for (TraceInfo t : traces) {
+    string exe = get_exec_path(t);
+
+    // Record date & runtime estimates
+    struct stat stat_version;
+    struct stat stat_data;
+    string version_file = traces_dir + "/" + t.first->d_name + "/version";
+    string data_file = traces_dir + "/" + t.first->d_name + "/data";
+    stat(version_file.c_str(), &stat_version);
+    stat(data_file.c_str(), &stat_data);
+    long int duration = stat_data.st_mtime - stat_version.st_mtime;
```

Comment:
This isn't a most robust way to get the duration. In particular if someone naively copies a trace from one place to another and then examines it with `rr ls`, the duration will be reported as zero, or something very small or even negative. I think that's a problem.

A foolproof way to get the duration would be to read the timestamps from the first and last trace records, but that can be slow for very large traces.

I think for now you should probably just remove the duration display so we can get this landed.